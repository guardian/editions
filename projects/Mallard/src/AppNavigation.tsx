import {
	CardStyleInterpolators,
	createStackNavigator,
} from '@react-navigation/stack';
import React from 'react';
import { Animated } from 'react-native';
import type { RootStackParamList } from './navigation/NavigationModels';
import { RouteNames } from './navigation/NavigationModels';
import { ArticleWrapper } from './navigation/navigators/article';
import { EditionsMenuScreen } from './screens/editions-menu-screen';
import { HomeScreen } from './screens/home-screen';
import { AuthSwitcherScreen } from './screens/identity-login-screen';
import { IssueScreen } from './screens/issue-screen';
import { OnboardingConsentScreen } from './screens/onboarding-screen';
import { SettingsScreen } from './screens/settings-screen';
import { AlreadySubscribedScreen } from './screens/settings/already-subscribed-screen';
import { ApiScreen } from './screens/settings/api-screen';
import { BetaProgrammeFAQsScreen } from './screens/settings/beta-programme-faqs';
import { CasSignInScreen } from './screens/settings/cas-sign-in-screen';
import { CreditsScreen } from './screens/settings/credits-screen';
import { EditionsScreen } from './screens/settings/editions-screen';
import { FAQScreen } from './screens/settings/faq-screen';
import {
	GdprConsentScreen,
	GdprConsentScreenForOnboarding,
} from './screens/settings/gdpr-consent-screen';
import { HelpScreen } from './screens/settings/help-screen';
import { ManageEditionsScreenWithHeader } from './screens/settings/manage-editions-screen';
import {
	PrivacyPolicyScreen,
	PrivacyPolicyScreenForOnboarding,
} from './screens/settings/privacy-policy-screen';
import { SubscriptionDetailsScreen } from './screens/settings/subscription-details-screen';
import { TermsAndConditionsScreen } from './screens/settings/terms-and-conditions-screen';

const Stack = createStackNavigator<RootStackParamList>();

const { multiply } = Animated;

const cardStyleInterpolator = (props: any) => {
	const translateX = multiply(
		props.current.progress.interpolate({
			inputRange: [0, 1],
			outputRange: [props.layouts.screen.width, 0],
			extrapolate: 'clamp',
		}),
		props.inverted,
	);

	return {
		// ...CardStyleInterpolators.forHorizontalIOS(props),
		cardStyle: {
			overflow: 'hidden',
			transform: [
				// Translation for the animation of the current card
				{
					translateX,
				},
			],
		},
		overlayStyle: {
			opacity: props.current.progress.interpolate({
				inputRange: [0, 1],
				outputRange: [0, 0.5],
				extrapolate: 'clamp',
			}),
		},
	};
};

const RootStack = () => {
	return (
		<Stack.Navigator
			initialRouteName={RouteNames.Home}
			screenOptions={{ gestureEnabled: false, headerShown: false }}
		>
			<Stack.Screen
				name={RouteNames.Issue}
				component={IssueScreen}
				options={{}}
			/>
			<Stack.Screen
				name={RouteNames.IssueList}
				component={HomeScreen}
				options={{
					gestureDirection: 'horizontal',
					cardStyle: { backgroundColor: 'transparent' },
					cardOverlayEnabled: true,
					cardStyleInterpolator,
				}}
			/>
			<Stack.Screen
				name={RouteNames.EditionsMenu}
				component={EditionsMenuScreen}
				options={{
					gestureDirection: 'horizontal-inverted',
					cardStyle: { backgroundColor: 'transparent' },
					cardOverlayEnabled: true,
					cardStyleInterpolator,
				}}
			/>
			<Stack.Screen
				name={RouteNames.Article}
				component={ArticleWrapper}
				options={{
					cardStyleInterpolator:
						CardStyleInterpolators.forModalPresentationIOS,
					gestureEnabled: true,
					gestureDirection: 'vertical',
				}}
			/>
			{/* <Stack.Screen
                name={RouteNames.Article}
                component={SlideCardJames}
                options={{
                    cardStyleInterpolator: props => {
                        return {
                            ...CardStyleInterpolators.forModalPresentationIOS(
                                props,
                            ),
                        }
                    },
                }}
            /> */}
			<Stack.Screen
				name={RouteNames.Settings}
				component={SettingsScreen}
			/>
			<Stack.Screen
				name={RouteNames.SignIn}
				component={AuthSwitcherScreen}
			/>
			{/* <Stack.Screen
                name={RouteNames.OnboardingStart}
            /> */}
			<Stack.Screen
				name={RouteNames.OnboardingConsent}
				component={OnboardingConsentScreen}
			/>
			<Stack.Screen
				name={RouteNames.OnboardingConsentInline}
				component={GdprConsentScreenForOnboarding}
			/>
			<Stack.Screen
				name={RouteNames.PrivacyPolicyInline}
				component={PrivacyPolicyScreenForOnboarding}
			/>
			<Stack.Screen
				name={RouteNames.AlreadySubscribed}
				component={AlreadySubscribedScreen}
			/>
			<Stack.Screen
				name={RouteNames.CasSignIn}
				component={CasSignInScreen}
			/>
			<Stack.Screen
				name={RouteNames.ManageEditionsSettings}
				component={ManageEditionsScreenWithHeader}
			/>
			{/** @TODO Fix the enable all button */}
			<Stack.Screen
				name={RouteNames.GdprConsent}
				component={GdprConsentScreen}
			/>
			<Stack.Screen
				name={RouteNames.PrivacyPolicy}
				component={PrivacyPolicyScreen}
			/>

			{/* ==== Inspect from here === */}
			<Stack.Screen
				name={RouteNames.Edition}
				component={EditionsScreen}
			/>
			<Stack.Screen
				name={RouteNames.TermsAndConditions}
				component={TermsAndConditionsScreen}
			/>
			<Stack.Screen
				name={RouteNames.BetaProgrammeFAQs}
				component={BetaProgrammeFAQsScreen}
			/>
			<Stack.Screen name={RouteNames.Help} component={HelpScreen} />
			<Stack.Screen name={RouteNames.Credits} component={CreditsScreen} />
			<Stack.Screen name={RouteNames.FAQ} component={FAQScreen} />
			<Stack.Screen
				name={RouteNames.SubscriptionDetails}
				component={SubscriptionDetailsScreen}
			/>
			<Stack.Screen name={RouteNames.Endpoints} component={ApiScreen} />
			{/* Turned off to remove Promise rejection error on Android */}
			{/* <Stack.Screen
                name={RouteNames.Storybook}
                component={StorybookScreen}
            /> */}
		</Stack.Navigator>
	);
};

export { RootStack };
