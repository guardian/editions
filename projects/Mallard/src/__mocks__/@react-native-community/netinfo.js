module.exports = {
	getCurrentState: jest.fn(() => Promise.resolve()),
	addListener: jest.fn(),
	removeListeners: jest.fn(),
	NetInfoStateType: {
		none: 'none',
		cellular: 'cellular',
		wifi: 'wifi',
		unknown: 'unknown',
	},
	fetch: jest.fn(() => Promise.resolve(true)),
};
