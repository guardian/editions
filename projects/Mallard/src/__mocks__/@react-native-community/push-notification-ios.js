module.exports = {
    FetchResult: {
        NoData: null,
    },
}
