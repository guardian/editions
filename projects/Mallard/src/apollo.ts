/**
 * Setup Apollo framework. At the moment we use Apollo Local State management
 * mechanism and don't fetch data from GraphQL endpoints.
 */
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { SETTINGS_RESOLVERS } from './helpers/settings/resolvers'
import { resolveWeather } from './helpers/location'

/**
 * Resolvers is what Apollo uses to get the value of field that has never been
 * cached before. For example if you have a query `{ foo @client }`, Apollo
 * looks for the field `foo` at the root of the cache. If not available, it'll
 * look for the resolver `Query.foo` (`Query` is just the "namespace" where
 * all queries start).
 *  *
 * Resolvers are functions, which can be asynchonous. They can gather data from
 * local storage (ex. settings), from local APIs (ex. permissions, geolocation),
 * from remote APIs (ex. REST API).
 *
 * Once a resolver returns a value, Apollo will refresh automatically
 * any component that was requiring this in a query via the `useQuery` function.
 */
const RESOLVERS = {
    Query: {
        ...SETTINGS_RESOLVERS,
        weather: resolveWeather,
    },
}

/**
 * The `Link` is the interface Apollo uses to reach GraphQL servers. Since we
 * don't fetch anything from GraphQL endpoints at the moment, we use a no-op
 * `Link` object. We *must* throw to prevent bad queries from carrying through.
 *
 * Apollo would try to use the `Link` if you have a query such as `{ foo }`
 * without the `@client` annotation. In our current use case, `@client` always
 * needs to be specified.
 */
const link = {
    split: () => {
        throw new Error('Not implemented. Did you forget to use `@client`?')
    },
    concat: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
    request: () => {
        throw new Error('not implemented Did you forget to use `@client`?')
    },
}

export const createApolloClient = () =>
    new ApolloClient({
        cache: new InMemoryCache(),
        link,
        resolvers: RESOLVERS,
    })
