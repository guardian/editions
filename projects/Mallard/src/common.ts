export * from '../../Apps/common/src/index';
