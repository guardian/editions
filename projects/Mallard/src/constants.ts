import { Platform } from 'react-native';
import Config from 'react-native-config';

const {
	ID_API_URL,
	MEMBERS_DATA_API_URL,
	ID_ACCESS_TOKEN,
	ITUNES_CONNECT_SHARED_SECRET,
	ANDROID_RELEASE_STREAM,
} = Config;

const FACEBOOK_CLIENT_ID = '528503751025697';

const AUTH_TTL = 86400000; // ms in a day

const CAS_ENDPOINT_URL = 'https://content-auth.guardian.co.uk/subs';

const GOOGLE_CLIENT_ID =
	Platform.OS === 'android'
		? __DEV__
			? '774465807556-5oq58e8pnfqn9sptsivri5kcs9mthef9'
			: '774465807556-cjat38acttpl7md4nfc4pfediouh7v97'
		: '774465807556-kgaj5an4pc4fmr3svp5nfpulekc1rl3n';

const LEGACY_SUBSCRIBER_ID_USER_DEFAULT_KEY = 'printSubscriberID';
const LEGACY_SUBSCRIBER_POSTCODE_USER_DEFAULT_KEY = 'printSubscriberPostcode';

const JOIN_BETA_LINK = Platform.select({
	ios: 'https://testflight.apple.com/join/O2EojUEl',
	android: 'https://play.google.com/apps/testing/com.guardian.editions',
});

export {
	CAS_ENDPOINT_URL,
	ID_API_URL,
	MEMBERS_DATA_API_URL,
	ID_ACCESS_TOKEN,
	FACEBOOK_CLIENT_ID,
	GOOGLE_CLIENT_ID,
	AUTH_TTL,
	LEGACY_SUBSCRIBER_ID_USER_DEFAULT_KEY,
	LEGACY_SUBSCRIBER_POSTCODE_USER_DEFAULT_KEY,
	ITUNES_CONNECT_SHARED_SECRET,
	ANDROID_RELEASE_STREAM,
	JOIN_BETA_LINK,
};
