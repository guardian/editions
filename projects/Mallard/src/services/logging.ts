// Logging Service that sends event logs to ELK
import NetInfo, { NetInfoStateType } from '@react-native-community/netinfo';
import type ApolloClient from 'apollo-client';
import gql from 'graphql-tag';
import { Platform } from 'react-native';
import Config from 'react-native-config';
import DeviceInfo from 'react-native-device-info';
import { getCASCode } from 'src/authentication/helpers';
import { isInBeta } from 'src/helpers/release-stream';
import type { GdprSwitchSetting } from 'src/helpers/settings';
import { defaultSettings } from 'src/helpers/settings/defaults';
import {
	iapReceiptCache,
	loggingQueueCache,
	userDataCache,
} from 'src/helpers/storage';
import {
	getDefaultEditionSlug,
	getSelectedEditionSlug,
} from 'src/hooks/use-edition-provider';
import type { MallardLogFormat } from '../../../Apps/common/src/logging';
import {
	Feature,
	Level,
	OS,
	ReleaseChannel,
} from '../../../Apps/common/src/logging';
import { AsyncQueue } from '../helpers/async-queue-cache';
import { errorService } from './errors';
import { remoteConfigService } from './remote-config';

const { LOGGING_API_KEY } = Config;
const ATTEMPTS_THEN_CLEAR = 10;
const MAX_NUM_OF_LOGS = 30;

interface LogParams {
	level: Level;
	message: string;
	optionalFields?: object;
}

type QueryData = { gdprAllowPerformance: GdprSwitchSetting };
const QUERY = gql('{ gdprAllowPerformance @client }');

const cropMessage = (message: string, maxLength: number): string => {
	return message.length > maxLength
		? `${message.slice(0, 6)}... (message cropped)`
		: message;
};

class Logging extends AsyncQueue {
	hasConsent: GdprSwitchSetting;
	numberOfAttempts: number;
	enabled: boolean;

	constructor() {
		super(loggingQueueCache);
		this.hasConsent = false;
		this.numberOfAttempts = 0;
		this.enabled = remoteConfigService.getBoolean('logging_enabled');
	}

	init(apolloClient: ApolloClient<object>) {
		apolloClient
			.watchQuery<QueryData>({ query: QUERY })
			.subscribe({
				next: (query) => {
					if (query.loading) return;
					this.hasConsent = query.data.gdprAllowPerformance;
				},
				error: (error) => {
					errorService.captureException(error);
				},
			});
	}

	async getExternalInfo() {
		const [
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		] = await Promise.all([
			NetInfo.fetch(),
			userDataCache.get(),
			getCASCode(),
			iapReceiptCache.get(),
		]);
		return {
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		};
	}

	async baseLog({
		level,
		message,
		...optionalFields
	}: LogParams): Promise<MallardLogFormat> {
		const {
			networkStatus,
			userData,
			casCode,
			iapReceipt,
		} = await this.getExternalInfo();

		// User Data and Subscription
		const userId = userData?.userDetails.id ?? '';
		const digitalSub =
			userData?.membershipData.contentAccess.digitalPack ?? false;
		const iAP = iapReceipt ? true : false;
		const selectedEdition = await getSelectedEditionSlug();
		const defaultEdition = await getDefaultEditionSlug();

		return {
			app: DeviceInfo.getBundleId(),
			version: DeviceInfo.getVersion(),
			buildNumber: DeviceInfo.getBuildNumber(),
			os: Platform.OS === 'ios' ? OS.IOS : OS.ANDROID,
			device: DeviceInfo.getDeviceId(),
			networkStatus: networkStatus
				? networkStatus.type
				: NetInfoStateType.unknown,
			selectedEdition: selectedEdition,
			defaultEdition: defaultEdition,
			release_channel: isInBeta()
				? ReleaseChannel.BETA
				: ReleaseChannel.RELEASE,
			timestamp: new Date(),
			level,
			message,
			deviceId: DeviceInfo.getUniqueId(),
			signedIn: userData ? true : false,
			userId,
			digitalSub,
			casCode,
			iAP,
			...optionalFields,
		};
	}

	async postLogToService(log: object[]): Promise<Response | Error> {
		const response = await fetch(defaultSettings.logging, {
			method: 'POST',
			headers: {
				Accept: 'application/json',
				'Content-Type': 'application/json',
				'x-api-key': LOGGING_API_KEY,
			},
			body: JSON.stringify(log),
		});
		if (response.status !== 200) {
			throw new Error(
				`Bad response from Logging Service - status: ${response.status}`,
			);
		}
		return response;
	}

	async postLogs() {
		try {
			const { isConnected } = await NetInfo.fetch();
			// Only attempt if we are connected, otherwise wait till next time
			if (isConnected) {
				const logsToPost = await this.getQueuedItems();

				if (logsToPost.length > 1) {
					await this.postLogToService(logsToPost);
					await this.clearItems();
					this.numberOfAttempts = 0;
				}
			}
		} catch (e) {
			if (this.numberOfAttempts >= ATTEMPTS_THEN_CLEAR) {
				await this.clearItems();
				this.numberOfAttempts = 0;
			} else {
				this.numberOfAttempts++;
			}
			errorService.captureException(e);
			return e;
		}
	}

	async log({ level, message, ...optionalFields }: LogParams) {
		try {
			if (!this.hasConsent) {
				return;
			}

			// limit max length of message we post to logging service
			const croppedMessage = cropMessage(message, 300);

			const currentLog = await this.baseLog({
				level,
				message: croppedMessage,
				...optionalFields,
			});

			return this.upsertQueuedItems([currentLog], MAX_NUM_OF_LOGS);
		} catch (e) {
			errorService.captureException(e);
			return e;
		}
	}

	async basicLogInfo({
		level,
		message,
		...optionalFields
	}: LogParams): Promise<MallardLogFormat> {
		return await this.baseLog({ level, message, ...optionalFields });
	}
}

const loggingService = new Logging();

export { Level, Feature, Logging, loggingService, cropMessage };
