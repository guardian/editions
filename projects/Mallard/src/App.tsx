// Welcome to the main entry point of the app.
//
// In this file, we'll be kicking off our app or storybook.

import { ApolloProvider } from '@apollo/react-hooks';
import AsyncStorage from '@react-native-community/async-storage';
import type ApolloClient from 'apollo-client';
import React from 'react';
import { AppState, StatusBar, StyleSheet, View } from 'react-native';
import SplashScreen from 'react-native-splash-screen';
import { eventEmitter } from 'src/helpers/event-emitter';
import { weatherHider } from 'src/helpers/weather-hider';
import {
	ConfigProvider,
	largeDeviceMemory,
} from 'src/hooks/use-config-provider';
import { NavPositionProvider } from 'src/hooks/use-nav-position';
import { setUserId } from 'src/services/ophan';
import { AppNavigation } from './AppNavigation';
import { AccessProvider } from './authentication/AccessContext';
import type { IdentityAuthData } from './authentication/authorizers/IdentityAuthorizer';
import type { AnyAttempt } from './authentication/lib/Attempt';
import { isValid } from './authentication/lib/Attempt';
import { BugButtonHandler } from './components/Button/BugButtonHandler';
import { ErrorBoundary } from './components/layout/ui/errors/error-boundary';
import { Modal, ModalRenderer } from './components/modal';
import { NetInfoDevOverlay } from './components/NetInfoDevOverlay';
import { NetInfoAutoToast } from './components/toast/net-info-auto-toast';
import { prepareAndDownloadTodaysIssue } from './download-edition/prepare-and-download-issue';
import { prepFileSystem } from './helpers/files';
import { nestProviders } from './helpers/provider';
import { pushDownloadFailsafe } from './helpers/push-download-failsafe';
import { isInBeta } from './helpers/release-stream';
import { newMobileProdStack } from './helpers/settings/defaults';
import { setApiUrl } from './helpers/settings/setters';
import { rollbackedApiUrlForBetaUsers } from './helpers/storage';
import { EditionProvider } from './hooks/use-edition-provider';
import { SettingsOverlayProvider } from './hooks/use-settings-overlay';
import { ToastProvider } from './hooks/use-toast';
import { pushNotificationRegistration } from './notifications/push-notifications';
import { DeprecateVersionModal } from './screens/deprecate-screen';
import { apolloClient } from './services/apollo-singleton';
import { errorService } from './services/errors';
import { loggingService } from './services/logging';
import { remoteConfigService } from './services/remote-config';

// Log Intitialisation
if (!__DEV__) {
	errorService.init(apolloClient);
}
loggingService.init(apolloClient);
remoteConfigService.init();

// --- SETUP OPERATIONS ---
pushNotificationRegistration();
prepFileSystem();

const styles = StyleSheet.create({
	appContainer: {
		flex: 1,
		backgroundColor: '#000',
	},
});

const persistenceKey = 'dev-nav-key-232asfdffgdfg1asdffgfdgfdga3413';

const isReactNavPersistenceError = (e: Error) =>
	__DEV__ && e.message.includes('There is no route defined for');

const WithProviders = nestProviders(
	Modal,
	ToastProvider,
	NavPositionProvider,
	ConfigProvider,
	EditionProvider,
	SettingsOverlayProvider,
);

const handleIdStatus = (attempt: AnyAttempt<IdentityAuthData>) =>
	setUserId(isValid(attempt) ? attempt.data.userDetails.id : null);

const shouldHavePushFailsafe = async (client: ApolloClient<object>) => {
	const largeRAM = await largeDeviceMemory();
	if (largeRAM) {
		pushDownloadFailsafe(client);
	}
};

const forceUpdateApiUrlIfBeta = async () => {
	// TODO - Remove this whole function before release to prod.
	// Previously we forced beta users to use a tmp url to point to the new stack, now that
	// we changed the fastly config we can rollback urls and eventually we can remove this full method.
	if (isInBeta()) {
		const rollbackedApi = await rollbackedApiUrlForBetaUsers.get();
		if (!rollbackedApi) {
			setApiUrl(apolloClient, newMobileProdStack);
			console.log('*** Beta: rolling back to default prod api url ***');
			await rollbackedApiUrlForBetaUsers.set(true);
		} else {
			console.log('*** Beta: api url for Beta users already updated ***');
		}
	}
};

export default class App extends React.Component {
	componentDidMount() {
		SplashScreen.hide();
		prepareAndDownloadTodaysIssue(apolloClient);
		shouldHavePushFailsafe(apolloClient);
		loggingService.postLogs();

		AppState.addEventListener('change', async (appState) => {
			if (appState === 'active') {
				prepareAndDownloadTodaysIssue(apolloClient);
				loggingService.postLogs();
				forceUpdateApiUrlIfBeta();
			}
		});

		{
			eventEmitter.on('editionCachesSet', () => {
				weatherHider(apolloClient);
			});
		}
	}

	async componentDidCatch(e: Error) {
		/**
		 * use an heuristic to check whether this is a react-nav error
		 * if it is then ditch our persistence and try to re-render
		 */
		if (isReactNavPersistenceError(e)) {
			await AsyncStorage.removeItem(persistenceKey);
			this.forceUpdate();
		}
	}
	/**
	 * When the component is mounted. This happens asynchronously and simply
	 * re-renders when we're good to go.
	 */
	render() {
		return (
			<ErrorBoundary>
				<ApolloProvider client={apolloClient}>
					<NetInfoDevOverlay>
						<WithProviders>
							<AccessProvider
								onIdentityStatusChange={handleIdStatus}
							>
								<StatusBar
									barStyle="light-content"
									backgroundColor="#041f4a"
								/>
								<View style={styles.appContainer}>
									<AppNavigation />
									<NetInfoAutoToast />
								</View>
								<ModalRenderer />
								<BugButtonHandler />
								<DeprecateVersionModal />
							</AccessProvider>
						</WithProviders>
					</NetInfoDevOverlay>
				</ApolloProvider>
			</ErrorBoundary>
		);
	}
}

/**
 * This needs to match what's found in your app_delegate.m and MainActivity.java.
 */

export const APP_NAME = 'Mallard';
