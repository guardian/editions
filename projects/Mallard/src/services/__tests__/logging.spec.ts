import { Logging, Level } from '../logging'
import MockDate from 'mockdate'
import { ReleaseChannel, OS } from '../../../../Apps/common/src/logging'
import { NetInfoStateType } from '@react-native-community/netinfo'

jest.mock('@react-native-community/netinfo', () => ({
    fetch: jest.fn(() => Promise.resolve({ isConnected: true })),
    NetInfoStateType: {
        unknown: 'unknown',
    },
}))
MockDate.set('2019-08-21')

const logFixture = [
    {
        timestamp: new Date(),
        level: Level.INFO,
        message: 'basic log',
        app: 'com.guardian.gce',
        version: '2.0',
        buildNumber: '678',
        release_channel: ReleaseChannel.BETA,
        os: OS.IOS,
        device: 'iPad4,1',
        networkStatus: NetInfoStateType.wifi,
        deviceId: '12345qwerty',
        signedIn: true,
        userId: null,
        digitalSub: false,
        casCode: null,
        iAP: false,
    },
]

describe('logging service', () => {
    describe('baseLog', () => {
        it('should return a log object that matches the snapshot', async () => {
            const loggingService = new Logging()
            loggingService.getExternalInfo = jest.fn().mockReturnValue({
                networkStatus: { type: 'wifi' },
                userData: {
                    userDetails: { id: 'testId' },
                    membershipData: {
                        contentAccess: { digitalPack: true },
                    },
                },
                casCode: 'QWERTYUIOP',
                iapReceipt: true,
            })
            const log = await loggingService.baseLog({
                level: Level.INFO,
                message: 'test log',
                optionalFields: { id: 'test' },
            })
            expect(log).toMatchSnapshot()
        })
        it('should return an object with default values if they are missing and match snapshot', async () => {
            const loggingService = new Logging()
            loggingService.getExternalInfo = jest.fn().mockReturnValue({
                networkStatus: null,
                userData: null,
                casCode: null,
                iapReceipt: null,
            })
            const log = await loggingService.baseLog({
                level: Level.INFO,
                message: 'test log',
                optionalFields: { id: 'test' },
            })
            expect(log).toMatchSnapshot()
        })
    })

    describe('log', () => {
        it('should have a successful post log', async () => {
            const loggingService = new Logging()
            loggingService.getExternalInfo = jest.fn().mockReturnValue({
                networkStatus: { type: 'wifi' },
                userData: {
                    userDetails: { id: 'testId' },
                    membershipData: {
                        contentAccess: { digitalPack: true },
                    },
                },
                casCode: 'QWERTYUIOP',
                iapReceipt: true,
            })
            loggingService.clearItems = jest.fn()
            loggingService.postLog = jest.fn()
            loggingService.hasConsent = true
            await loggingService.log({ level: Level.INFO, message: 'test' })
            expect(loggingService.postLog).toHaveBeenCalled()
            expect(loggingService.clearItems).toHaveBeenCalled()
        })
        it('should not post a log if there is no consent', async () => {
            const loggingService = new Logging()
            loggingService.getExternalInfo = jest.fn().mockReturnValue({
                networkStatus: { type: 'wifi' },
                userData: {
                    userDetails: { id: 'testId' },
                    membershipData: {
                        contentAccess: { digitalPack: true },
                    },
                },
                casCode: 'QWERTYUIOP',
                iapReceipt: true,
            })
            loggingService.clearItems = jest.fn()
            loggingService.postLog = jest.fn()

            await loggingService.log({ level: Level.INFO, message: 'test' })
            expect(loggingService.postLog).not.toHaveBeenCalled()
            expect(loggingService.clearItems).not.toHaveBeenCalled()
        })
    })

    describe('postLog', () => {
        it('should increase the number of attempts if postLog fails', async () => {
            const loggingService = new Logging()
            loggingService.postLogToService = jest
                .fn()
                .mockImplementation(() => {
                    throw new Error('unable to post')
                })

            try {
                await loggingService.postLog(logFixture)
            } catch {
                expect(loggingService.numberOfAttempts).toEqual(1)
            }
        })
        it('should reset the number of attempts on a successful post', async () => {
            const loggingService = new Logging()
            loggingService.postLogToService = jest
                .fn()
                .mockReturnValue(Promise.resolve(true))
            await loggingService.postLog(logFixture)
            expect(loggingService.numberOfAttempts).toEqual(0)
        })
        it('should call clearLogs if the threshold is reached when there is an error', async () => {
            const loggingService = new Logging()
            loggingService.numberOfAttempts = 10
            loggingService.postLogToService = jest
                .fn()
                .mockImplementation(() => {
                    throw new Error('unable to post')
                })
            loggingService.clearItems = jest.fn()

            try {
                await loggingService.postLog(logFixture)
            } catch {
                expect(loggingService.numberOfAttempts).toEqual(0)
                expect(loggingService.clearItems).toHaveBeenCalled()
            }
        })
    })
})
