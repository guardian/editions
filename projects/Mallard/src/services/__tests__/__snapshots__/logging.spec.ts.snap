// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logging service baseLog should return a log object that matches the snapshot 1`] = `
Object {
  "app": "com.guardian.gce",
  "buildNumber": "666",
  "casCode": "QWERTYUIOP",
  "defaultEdition": null,
  "device": "iPad4,1",
  "deviceId": "FCDBD8EF-62FC-4ECB-B2F5-92C9E79AC7F9",
  "digitalSub": true,
  "iAP": true,
  "level": "INFO",
  "message": "test log",
  "networkStatus": "wifi",
  "optionalFields": Object {
    "id": "test",
  },
  "os": "ios",
  "release_channel": "BETA",
  "selectedEdition": "daily-edition",
  "signedIn": true,
  "timestamp": 2019-08-21T00:00:00.000Z,
  "userId": "testId",
  "version": "3.1",
}
`;

exports[`logging service baseLog should return an object with default values if they are missing and match snapshot 1`] = `
Object {
  "app": "com.guardian.gce",
  "buildNumber": "666",
  "casCode": null,
  "defaultEdition": null,
  "device": "iPad4,1",
  "deviceId": "FCDBD8EF-62FC-4ECB-B2F5-92C9E79AC7F9",
  "digitalSub": false,
  "iAP": false,
  "level": "INFO",
  "message": "test log",
  "networkStatus": "unknown",
  "optionalFields": Object {
    "id": "test",
  },
  "os": "ios",
  "release_channel": "BETA",
  "selectedEdition": "daily-edition",
  "signedIn": false,
  "timestamp": 2019-08-21T00:00:00.000Z,
  "userId": "",
  "version": "3.1",
}
`;
